File: python_code/highlevel/heap/PriorityQueue.py
# 抖码算法，让算法学习变得简单有趣
# 作者：老汤
from highlevel.heap.MaxHeap import MaxHeap


class PriorityQueue:
    def __init__(self, __comp):
        self.maxHeap = MaxHeap(__comp)

    def getSize(self):
        return self.maxHeap.size()

    def isEmpty(self):
        return self.maxHeap.isEmpty()

    def enqueue(self, e):
        self.maxHeap.add(e)

    def dequeue(self):
        return self.maxHeap.removeMax()

    def getFront(self):
        return self.maxHeap.findMax()

File: cpp_code/highlevel/heap/PriorityQueue.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include "./MaxHeap.cpp"

template <typename E, typename Compare>
class PriorityQueue {
private:
    MaxHeap<E, Compare>* maxHeap;

public:
    PriorityQueue() {
        this->maxHeap = new MaxHeap<E, Compare>();
    }

    PriorityQueue(Compare __comp) {
        this->maxHeap = new MaxHeap<E, Compare>(__comp);
    }

    int getSize() {
        return maxHeap->size();
    }

    bool sEmpty() {
        return maxHeap->isEmpty();
    }

    void enqueue(E e) {
        maxHeap->add(e);
    }

    E dequeue() {
        return maxHeap->removeMax();
    }

    E getFront() {
        return maxHeap->findMax();
    }
};